server/src/services/unlock-protocol.ts

```diff
@@ -0,0 +1,59 @@
+import {UnlockTask, Services} from '../types';
+import { getABI, mintToken, checkAddress, checkHasToken } from '../eth/helpers';
+import { Contract, getDefaultProvider } from 'ethers';
+import { UnlockProtocol } from '../eth/UnlockProtocol';
+import { hasToken, finishTaskWithErrors, finishTask, setInProcessTask, setPendingTask } from '../db';
+
+const ABI = getABI('UnlockProtocol');
+
+const eventID = 80;
+
+export async function processUnlockTask(task :UnlockTask){
+  if(task.name !== Services.unlockProtocol) return;
+
+  // Check that the accountAddress is a valid address
+  if(await checkAddress(task.task_data.accountAddress) == null) {
+    finishTaskWithErrors('Invalid account address', task.id);
+    return;
+  }
+
+  // Check that the lockAddress is a valid address
+  if(await checkAddress(task.task_data.lockAddress) == null) {
+    finishTaskWithErrors('Invalid lock address', task.id);
+    return;
+  }
+
+  // Check if the address has a token
+  if(await hasToken(task)){
+    finishTaskWithErrors("Token already minted or in process", task.id);
+    return;
+  }
+
+  // Instantiate the unlock contract
+  let unlock = new Contract(task.task_data.lockAddress, ABI, getDefaultProvider()) as UnlockProtocol;
+  // Check if the address has a valid key 
+  const hasValidKey = await unlock.functions.getHasValidKey(task.task_data.accountAddress);
+
+  if(!hasValidKey){
+    return;
+  }
+
+  const has_token = await checkHasToken(eventID, task.task_data.accountAddress);
+  if (has_token) {
+    finishTaskWithErrors("Address already has token in blockchain", task.id);
+    return;
+  }
+
+  await setInProcessTask(eventID);
```

Comment:
@nahu90 thanks for the rework, but I still have doubts regarding the implementation.
Why are we updating the status here instead of line 13? We are making several operations before getting to this line.
I would update the status to `IN_PROCESS` right after checking that the task is type Unlock. If the token is not in the Unlock contract (line 37), I would update the task back to `PENDING`